README.md
# herf-time

[![Circle CI](https://circleci.com/gh/smurphy8/herf-time.svg?style=svg)](https://circleci.com/gh/smurphy8/herf-time)

Haskell's time library is nice but I wanted to be able to do some of the operations in
the [Kerf](https://github.com/kevinlawler/kerf) programming language.


By creating a set of type classes: **HerfedTime** , **ToUTCHerfTime**, **FromUTCHerfTime**
I think there is a nice mechanism to do just that.

any instance of these 3 classes should satisfy the law:

* **Path Independant** on all addtion 
* ``` addX a X  == unherf (addX (herf a) X )```
  * where: ``` addX := {addYear , addMonth, addWeek ...}```


# Tutorial 
[Herf Tutorial](http://blog.plowtech.net/posts/herf-time.html)

## Usage


### Add Intervals
``` haskell
>>> date 2016 01 01 `add` hour 3 `add` week 16 `add` month 3 :: UTCTime
UTCHerfTime 2016-07-22 03:00:00 UTC
```

### Subtract Intervals
``` haskell

>>> date 2016 01 01 `add` hour (-3) `add` week (-16) `add` month (-3) :: UTCTime
UTCHerfTime 2015-06-10 21:00:00 UTC
```

### Represent Time in Multiple Ways
``` haskell

>>> dateTime 2016 01 01 01 23 01 `add` hour 3 `add` week 16 `add` month 3 :: UTCTime
UTCHerfTime 2016-07-22 04:23:01 UTC
>>> dateTimePico 2016 01 01 01 23 01 01 `add` hour 3 `add` week 16 `add` month 3 :: UTCTime
UTCHerfTime 2016-07-22 04:23:01.000000000001 UTC
```
### Get Times in any HerfedTime format  (UTC for example)
``` haskell
>>> date 2016 01 01 `add` hour 3 `add` week 16 `add` month 3   :: UTCTime
2016-07-22 03:00:00 UTC
```

### Use HerfTime.ZonedTime to convert easily between times
(reherf $ ( dateTime 2016 01 01 01 01 01 :: HerfZonedTime "CST")) :: HerfZonedTime "PST"
2015-12-31T23:01:01:PST


### Some fancier examples (Using HerfTime.ZonedTime)


``` haskell
λ> zt <- getZonedTime
λ> herfShow zt
"2016-01-21T11:29:05:CST"
λ> reherfz zt :: HerfZonedTime "PST"
2016-01-21T09:29:05:PST
λ> reherfz zt :: HerfZonedTime "+0600"
2016-01-21T23:29:05:+0600
λ> reherfz zt :: HerfZonedTime "CST"
2016-01-21T11:29:05:CST
λ> (reherfz zt) `add` month 2 :: HerfZonedTime "CST"
2016-03-21T11:29:05:CST

-- What time will it be exactly 3 months from now in California?
λ> (reherfz zt) `add` month 2 :: HerfZonedTime "PST"
2016-03-21T09:29:05:PST
```




```
cabal configure --enable-tests && cabal build && cabal test
```




ChangeLog.md
# Revision history for kerf-time
## 0.3.1
* fix `time` dependency to add ParseTime for 1.10 + still backwards compatabile
## 0.3.0
* upgrade to ghc 8.2.2, lts-11.7

## 0.2.2
* Added HerfedTime constraint to herfShow
* Restricted exports in HerfTime module
* exported Time through both HerfTime module and ZonedTime module (to cut down on necessary exports)

## 0.2.1
Started using the changelog

## 0.2.0
Fixed module exposed problem
## 0.1.0  -- Deprecated

* First version. Released on an unsuspecting world.


test/Spec.hs
import           Test.DocTest
main :: IO ()
main = doctest ["-isrc", "src/HerfTime.hs","src/HerfTime/ZonedTime.hs"]




test/KerfTimeSpec.hs
module KerfTimeSpec (tests) where

import KerfTime

import Test.Tasty
import Test.Tasty.QuickCheck as QC
import Test.Tasty.HUnit

import Data.List
import Data.Ord

tests :: TestTree
tests = testGroup "Tests" [properties, unitTests]

properties :: TestTree
properties = testGroup "Properties" [qcProps]

qcProps = testGroup "(checked by QuickCheck)"
  [ QC.testProperty "sort == sort . reverse" (
      \list -> sort (list :: [Int]) == sort (reverse list))
  , QC.testProperty "Fermat's little theorem" (
      \x -> ((x :: Integer)^7 - x) `mod` 7 == 0)
  -- the following property does not hold
  , QC.testProperty "Fermat's last theorem" (
      \x y z n ->
        (n :: Integer) >= 3 QC.==> x^n + y^n /= (z^n :: Integer))
  ]

unitTests = testGroup "Unit tests"
  [ testCase "List comparison (different length)" (
      [1, 2, 3] `compare` [1,2] @?= GT)

  -- the following test does not hold
  , testCase "List comparison (same length)" (
      [1, 2, 3] `compare` [1,2,2] @?= LT)
  ]

src/KerfTime.hs
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
module KerfTime  where

import           Data.Time

-- | --------------------------------------------------
-- | Kerf Time Constructors
-- | --------------------------------------------------


-- | Usage: Add Intervals of different amounts
-- >>> date 2016 01 01 `add` hour 3 `add` week 16 `add` month 3 :: UTCKerfTime
-- UTCKerfTime 2016-07-22 03:00:00 UTC

-- | Represent Time in a few different ways:
-- >>> dateTime 2016 01 01 01 23 01 `add` (hour 3) `add` (week 16) `add` (month 3) :: UTCKerfTime
-- UTCKerfTime 2016-07-22 04:23:01 UTC
-- >>> dateTimePico 2016 01 01 01 23 01 01 `add` (hour 3) `add` (week 16) `add` (month 3) :: UTCKerfTime
-- UTCKerfTime 2016-07-22 04:23:01.000000000001 UTC


-- | Usage: Use negative signs to subtract
-- >>> date 2016 01 01 `add` hour (-3) `add` week (-16) `add` month (-3) :: UTCKerfTime
-- UTCKerfTime 2015-06-10 21:00:00 UTC


year :: Integer -> KerfYear
year = KerfYear

month :: Integer -> KerfMonth
month = KerfMonth

week :: Integer -> KerfWeek
week = KerfWeek

day :: Integer -> KerfDay
day = KerfDay

hour :: Integer -> KerfHour
hour = KerfHour

minute :: Integer -> KerfMin
minute = KerfMin

second :: Integer -> KerfSec
second = KerfSec

pico :: Integer -> KerfPico
pico = KerfPico

-- | >>> date 1 1 1 :: UTCKerfTime
-- UTCKerfTime 0001-01-01 00:00:00 UTC

dateKerf :: Integer -> Integer-> Integer-> UTCKerfTime
dateKerf y m d = UTCKerfTime $  UTCTime dayPart timePart
  where
    dayPart = fromGregorian y (fromInteger m) (fromInteger d)
    timePart = 0


-- | Time only, you can't just add a diff time to a date so we get a diff time back
-- >>> time 1 1 1
-- 3661s
time :: Integer -> Integer -> Integer -> DiffTime
time h m s = secondsToDiffTime ( convertedHours  +
                                 convertedMinutes +
                                 convertedSeconds)
  where
    convertedHours = h * 3600
    convertedMinutes = m * 60
    convertedSeconds = s


timePico :: Integer -> Integer -> Integer -> Integer -> DiffTime
timePico h m s p = picoTime +
                   (time h m s)
 where
   picoTime = picosecondsToDiffTime p

dateTimeKerf :: Integer -> Integer -> Integer ->
                Integer -> Integer -> Integer -> UTCKerfTime
dateTimeKerf y m d h i s = UTCKerfTime $ UTCTime dayPart timePart
  where
    dayPart = fromGregorian y (fromInteger m) (fromInteger d)
    timePart = time h i s

dateTimePicoKerf :: Integer -> Integer -> Integer
                 -> Integer -> Integer  -> Integer -> Integer
                 -> UTCKerfTime
dateTimePicoKerf y m d h i s p = UTCKerfTime $ UTCTime dayPart timePart
  where
    dayPart = fromGregorian y (fromInteger m) (fromInteger d)
    timePart = timePico h i s p


-- --------------------------------------------------

newtype UTCKerfTime = UTCKerfTime UTCTime
  deriving (Eq,Ord,Show)

class (Eq a, Ord a) => ToUTCKerfTime a where
  kerf :: a -> UTCKerfTime

class FromUTCKerfTime a where
  unkerf ::  UTCKerfTime -> a

class (ToUTCKerfTime a, FromUTCKerfTime a) => KerfedTime a where
  addYear :: a -> KerfYear -> a
  addMonth :: a -> KerfMonth -> a
  addWeek :: a -> KerfWeek -> a
  addDay :: a -> KerfDay -> a
  addHour :: a -> KerfHour -> a
  addMinute :: a -> KerfMin -> a
  addSecond :: a -> KerfSec -> a
  addPicosecond :: a -> KerfPico -> a
  date :: KerfYear -> KerfMonth -> KerfDay -> a
  dateTime :: KerfYear -> KerfMonth -> KerfDay -> KerfHour -> KerfMin -> KerfSec ->  a
  dateTimePico :: KerfYear -> KerfMonth -> KerfDay -> KerfHour -> KerfMin -> KerfSec -> KerfPico ->  a



class KerfAdd a where
  add :: (KerfedTime t) => t -> a -> t

instance KerfAdd KerfYear where
  add =  addYear

instance KerfAdd KerfMonth where
  add =  addMonth

instance KerfAdd KerfWeek where
  add =  addWeek

instance KerfAdd KerfDay where
  add =  addDay

instance KerfAdd KerfHour where
  add =  addHour

instance KerfAdd KerfMin where
  add =  addMinute

instance KerfAdd KerfSec where
  add =  addSecond

instance KerfAdd KerfPico where
  add =  addPicosecond


newtype KerfYear   = KerfYear  Integer
  deriving (Num,Eq,Ord,Show)
newtype KerfMonth  = KerfMonth Integer
  deriving (Num,Eq,Ord,Show)
newtype KerfWeek = KerfWeek Integer
  deriving (Num,Eq,Ord,Show)
newtype KerfDay    = KerfDay   Integer
  deriving (Num,Eq,Ord,Show)
newtype KerfMin    = KerfMin   Integer
  deriving (Num,Eq,Ord,Show)
newtype KerfHour   = KerfHour  Integer
  deriving (Num,Eq,Ord,Show)
newtype KerfSec    = KerfSec   Integer
  deriving (Num,Eq,Ord,Show)
newtype KerfPico   = KerfPico  Integer -- Kerf uses nano but whatev
  deriving (Num, Eq, Ord, Show)


-- | NominalKerf time will obey the laws of KerfedTime by default
instance ToUTCKerfTime UTCKerfTime where
  kerf = id

instance FromUTCKerfTime UTCKerfTime where
  unkerf  = id

instance KerfedTime UTCKerfTime where
  addYear (UTCKerfTime k) y = UTCKerfTime $ addYear k y
  addMonth (UTCKerfTime k) m = UTCKerfTime $ addMonth k m
  addWeek (UTCKerfTime k) w = UTCKerfTime $ addWeek k w
  addDay (UTCKerfTime k) d = UTCKerfTime $ addDay k d
  addHour (UTCKerfTime k) h = UTCKerfTime $ addHour k h
  addMinute (UTCKerfTime k) i = UTCKerfTime $ addMinute k i
  addSecond (UTCKerfTime k) s = UTCKerfTime $ addSecond k s
  addPicosecond (UTCKerfTime k) p = UTCKerfTime $ addPicosecond k p
  date (KerfYear y) (KerfMonth m) (KerfDay d) = dateKerf y m d
  dateTime (KerfYear y) (KerfMonth m) (KerfDay d) (KerfHour h) (KerfMin i) (KerfSec s ) = dateTimeKerf y m d h i s
  dateTimePico (KerfYear y) (KerfMonth m) (KerfDay d) (KerfHour h) (KerfMin i) (KerfSec s ) (KerfPico p ) = dateTimePicoKerf y m d h i s p
-- | UTCTime is the underlying and most important KerfTime thing
instance ToUTCKerfTime UTCTime where
  kerf = UTCKerfTime

instance FromUTCKerfTime UTCTime where
  unkerf (UTCKerfTime u) = u

-- | Get Times in any viable format (UTC for example)
-- >>> unkerf $ date 2016 01 01 `add` hour 3 `add` week 16 `add` month 3   :: UTCTime
-- 2016-07-22 03:00:00 UTC

instance KerfedTime UTCTime where
  addYear (UTCTime d t) (KerfYear y) = UTCTime (addGregorianYearsRollOver y d) t
  addMonth (UTCTime d t) (KerfMonth m) = UTCTime (addGregorianMonthsRollOver m d) t
  addWeek  (UTCTime d t) (KerfWeek w) = UTCTime (addDays (7*w) d) t
  addDay (UTCTime d t) (KerfDay ds) = UTCTime (addDays ds d) t
  addHour u (KerfHour h) = addUTCTime (fromIntegral $ h*3600) u
  addMinute u (KerfMin i) = addUTCTime (fromIntegral $ i*60) u
  addSecond u (KerfSec s) = addUTCTime (fromIntegral s) u
  addPicosecond u (KerfPico p) = addUTCTime (toNominal p) u
    where
      toNominal = fromRational . toRational .  picosecondsToDiffTime

  date (KerfYear y) (KerfMonth m) (KerfDay d) = unkerf $ dateKerf y m d
  dateTime (KerfYear y) (KerfMonth m) (KerfDay d) (KerfHour h) (KerfMin i) (KerfSec s ) = unkerf $ dateTimeKerf y m d h i s
  dateTimePico (KerfYear y) (KerfMonth m) (KerfDay d) (KerfHour h) (KerfMin i) (KerfSec s ) (KerfPico p ) = unkerf $ dateTimePicoKerf y m d h i s p


















--------------------------------------------------
-- Retrieval
--------------------------------------------------
-- | getYear someUtcTime -> 2016y
getYear :: UTCTime -> Integer
getYear incomingTime = case incomingTime of
   (UTCTime d _) -> let (year',_ ,_ ) = toGregorian d
                     in year'


-- | getMonth someUtcTime -> 10m
getMonth :: UTCTime -> Integer
getMonth incomingTime = case incomingTime of
   (UTCTime d _) -> let (_ , month' ,_ ) = toGregorian d
                    in  fromIntegral month'


-- | getDay someUtcTime -> 31d
getDay :: UTCTime -> Integer
getDay incomingTime = case incomingTime of
   (UTCTime d _) -> let (_ , _ ,day' ) = toGregorian d
                    in  fromIntegral day'


-- | Get all date parts together
getDateParts :: UTCTime -> (Integer,Integer,Integer)
getDateParts (UTCTime d _)  = (year',fromIntegral month',fromIntegral day')
  where
    (year',month',day') = toGregorian d

-- | getHour someUtcTime -> 1h
getHour :: UTCTime -> Integer
getHour (UTCTime _ t) = floor (t / 3600)

-- | getMin someUtcTime -> 37i
getMin :: UTCTime -> Integer
getMin u@(UTCTime _ t) = div remainingSeconds 60
  where
    timeInSeconds = floor t
    remainingSeconds = timeInSeconds - secondsInHours
    secondsInHours = 3600 * (getHour u)


-- | getSeconds someUtcTime -> 1s
getSeconds :: UTCTime -> Integer
getSeconds u@(UTCTime _ t) = remainingSeconds
  where
    timeInSeconds = floor t
    remainingSeconds = timeInSeconds - secondsInHours - secondsInMinutes
    secondsInHours = 3600 * (getHour u)
    secondsInMinutes = 60 * (getMin u)

getPicoseconds :: UTCTime -> Integer
getPicoseconds u@(UTCTime _ t) = round $ remainingPico *
                                  (fromRational (10^(12 :: Integer)))
  where
    remainingPico = t - (fromIntegral $ secondsInHours - secondsInMinutes - seconds')
    secondsInHours = 3600 * (getHour u)
    secondsInMinutes = 60 * (getMin u)
    seconds' =  getSeconds u


src/HerfTime/ZonedTime.hs
{-# LANGUAGE DataKinds                  #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE KindSignatures             #-}
{-# LANGUAGE ScopedTypeVariables        #-}
{-# LANGUAGE CPP                        #-}
{- |
Module      : HerfTime.LocalTime
Description : LocalTime interpreter for HerfedTime Class
Copyright   : Plow Technologies LLC
License     : MIT License

Maintainer  : Scott Murphy


The basic idea is to use the regular local time implementation available in the time library
but with a few nice wrappers.

For HerfTime to work nicely the information about what makes a time different should reside in the type.

in the case of zoned time, this is the time zone:

HerfZonedTime knows about these time zones by name:

RFC 822 sec. 5: "UT", "GMT", "EST", "EDT", "CST", "CDT", "MST", "MDT", "PST", "PDT".



| -}


module HerfTime.ZonedTime ( HerfZonedTime
                          , toZonedTime
                          , fromZonedTime
                          , addTimeZone
                          , herfz
                          , reherfz
                          , module HerfTime
                          , module Data.Time) where

import           Control.Applicative ((<|>))
import           Data.Maybe          (fromMaybe)
import           Data.Proxy
import           Data.Time
import           GHC.TypeLits
import           HerfTime
#if MIN_VERSION_time(1,9,0) && !MIN_VERSION_time(1,10,0)
import           Data.Time.Format.Internal (ParseTime(..))  
#endif


-- | Zoned Time always has an extra parameter to convert into a fixed time

newtype HerfZonedTime (z::Symbol)  = HerfZonedTime {_unherfZonedTime :: ZonedTime}
#if MIN_VERSION_time(1,9,0) && !MIN_VERSION_time(1,10,0)
  deriving (FormatTime)
#else
  deriving (FormatTime, ParseTime)
#endif

#if MIN_VERSION_time(1,9,0) && !MIN_VERSION_time(1,10,0)
instance ParseTime (HerfZonedTime z) where
  buildTime locale fcs =
    case buildTime locale fcs of
      Nothing -> Nothing
      Just zonedTime -> Just $ HerfZonedTime {_unherfZonedTime = zonedTime}
  parseTimeSpecifier _ = parseTimeSpecifier (undefined :: m (ZonedTime))
#endif

instance (KnownSymbol z) => Show (HerfZonedTime z) where
  show = herfShow

-- | Add Time Zone is different than converting
-- It takes a UTCTime and just slaps on the timezone in the ZonedTime field
-- This is exactly what you want when you are first creating a time
-- and almost never what you want after!


addTimeZone :: forall z. (KnownSymbol z) =>  UTCTime -> HerfZonedTime z
addTimeZone (UTCTime day' diffTime') = HerfZonedTime zonedTime
  where
    zonedTime = ZonedTime localTime timeZone
    localTime = LocalTime day' (timeToTimeOfDay diffTime')
    pz = Proxy  :: Proxy z
    v = symbolVal pz
    timeZone :: TimeZone
    timeZone = case (parseTimeM True defaultTimeLocale "%Z" v ) of
                  Just t -> t
                  Nothing -> fromMaybe (error $ "time zone broken " ++ v) (parseTimeM True defaultTimeLocale "%z" v)



-- | Helper function to convert UTCTime to HerfZoned time
toZonedTime :: forall z . (KnownSymbol z) =>  UTCTime -> HerfZonedTime z
toZonedTime time' = HerfZonedTime $ utcToZonedTime (fromMaybe utc . tz $ directSymbolVal) time'
  where
    pz = Proxy  :: Proxy z
    directSymbolVal = symbolVal pz
    tz :: String -> Maybe TimeZone
    tz v = (parseTimeM True defaultTimeLocale "%Z" v ) <|>
           (parseTimeM True defaultTimeLocale "%z" v)


-- | Getting back from HerfZonedTime to UTCTime
fromZonedTime :: forall z . (KnownSymbol z) => HerfZonedTime z -> UTCTime
fromZonedTime (HerfZonedTime time') = zonedTimeToUTC time'



instance (KnownSymbol z) =>  ToUTCHerfTime (HerfZonedTime z) where
  herf = herf . fromZonedTime

instance (KnownSymbol z) => FromUTCHerfTime (HerfZonedTime z) where
  unherf = toZonedTime . unherf

-- |
-- >>> :set -XDataKinds
-- >>> (reherf $ ( dateTime 2016 01 01 01 01 01 :: HerfZonedTime "CST")) :: HerfZonedTime "PST"
-- 2015-12-31T23:01:01:PST

instance (KnownSymbol z) => HerfedTime (HerfZonedTime z) where
  addYear a y  = unherf $ herf a `add` y
  addMonth a m  = unherf $ herf a `add` m
  addWeek a w  = unherf $ herf a `add` w
  addDay a d  = unherf $ herf a `add` d
  addHour a h  = unherf $ herf a `add` h
  addMinute a i  = unherf $ herf a `add` i
  addSecond a s  = unherf $ herf a `add` s
  addPicosecond a p  = unherf $ herf a `add` p
  date y m d = toZonedTime $ date y m d
  dateTime y m d h i s = addTimeZone $ dateTime y m d h i s
  dateTimePico y m d h i s p = addTimeZone $ dateTimePico y m d h i s p


-- | Don't want an orphan on ZonedTime so I made herfz
herfz :: ZonedTime -> UTCHerfTime
herfz = herf . zonedTimeToUTC


-- | like 'reherf' but for zoned time (which doesn't have a direct HerfedTime instance)
reherfz :: FromUTCHerfTime b => ZonedTime -> b
reherfz = unherf . herfz




src/HerfTime.hs
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-|
The Herf Time package is loosely based on the Kerf programming language's approach to time.
<https://github.com/kevinlawler/kerf>

Where it made sense to do things differently it does.

The fundamental engine is the typeclass "HerfTime" -}


module HerfTime  ( module Data.Time
                 , herfShow
                 , reherf
                 , getYear
                 , getMonth
                 , getDay
                 , getDateParts
                 , getHour
                 , getMin
                 , getSeconds
                 , getPicoseconds
                 , UTCHerfTime
                 , HerfedTime(..)
                 , ToUTCHerfTime(..)
                 , FromUTCHerfTime(..)
                 , HerfAdd(..)
                 , HerfYear
                 , HerfMonth
                 , HerfWeek
                 , HerfDay
                 , HerfHour
                 , HerfMin
                 , HerfSec
                 , HerfPico
                 , year
                 , month
                 , week
                 , day
                 , hour
                 , minute
                 , second
                 , pico



                 ) where

import           Data.Time




-- | Add Intervals of different amounts

-- | >>> date 2016 01 01 `add` hour 3 `add` week 16 `add` month 3 :: UTCHerfTime
-- UTCHerfTime 2016-07-22 03:00:00 UTC

-- | Represent Time in a few different ways

-- | >>> dateTime 2016 01 01 01 23 01 `add` (hour 3) `add` (week 16) `add` (month 3) :: UTCHerfTime
-- UTCHerfTime 2016-07-22 04:23:01 UTC
-- >>> dateTimePico 2016 01 01 01 23 01 01 `add` (hour 3) `add` (week 16) `add` (month 3) :: UTCHerfTime
-- UTCHerfTime 2016-07-22 04:23:01.000000000001 UTC


-- | Use negative signs to subtract

-- | >>> date 2016 01 01 `add` hour (-3) `add` week (-16) `add` month (-3) :: UTCHerfTime
-- UTCHerfTime 2015-06-10 21:00:00 UTC

-- | Interface functions following Kerf
-- y
year :: Integer -> HerfYear
year = HerfYear

-- | m
month :: Integer -> HerfMonth
month = HerfMonth

-- | w
week :: Integer -> HerfWeek
week = HerfWeek

-- | d
day :: Integer -> HerfDay
day = HerfDay

-- |h
hour :: Integer -> HerfHour
hour = HerfHour

-- |i
minute :: Integer -> HerfMin
minute = HerfMin

-- | s
second :: Integer -> HerfSec
second = HerfSec

-- | p
pico :: Integer -> HerfPico
pico = HerfPico



-- | Display herf times in a pre formatted way
herfShow :: (HerfedTime t, FormatTime t) => t -> String
herfShow = formatTime defaultTimeLocale (iso8601DateFormat (Just "%H:%M:%S:%Z") )  -- i.e. YYYY-MM-DDTHH:MM:SS

-- | >>> date 1 1 1 :: UTCHerfTime
-- UTCHerfTime 0001-01-01 00:00:00 UTC

dateHerf :: Integer -> Integer-> Integer-> UTCHerfTime
dateHerf y m d = UTCHerfTime $  UTCTime dayPart timePart
  where
    dayPart = fromGregorian y (fromInteger m) (fromInteger d)
    timePart = 0


-- | Time only, you can't just add a diff time to a date so we get a diff time back
-- >>> time 1 1 1
-- 3661s
time :: Integer -> Integer -> Integer -> DiffTime
time h m s = secondsToDiffTime ( convertedHours  +
                                 convertedMinutes +
                                 convertedSeconds)
  where
    convertedHours = h * 3600
    convertedMinutes = m * 60
    convertedSeconds = s


timePico :: Integer -> Integer -> Integer -> Integer -> DiffTime
timePico h m s p = picoTime +
                   (time h m s)
 where
   picoTime = picosecondsToDiffTime p

dateTimeHerf :: Integer -> Integer -> Integer ->
                Integer -> Integer -> Integer -> UTCHerfTime
dateTimeHerf y m d h i s = UTCHerfTime $ UTCTime dayPart timePart
  where
    dayPart = fromGregorian y (fromInteger m) (fromInteger d)
    timePart = time h i s

dateTimePicoHerf :: Integer -> Integer -> Integer
                 -> Integer -> Integer  -> Integer -> Integer
                 -> UTCHerfTime
dateTimePicoHerf y m d h i s p = UTCHerfTime $ UTCTime dayPart timePart
  where
    dayPart = fromGregorian y (fromInteger m) (fromInteger d)
    timePart = timePico h i s p


-- --------------------------------------------------

-- | Below are the classes that make up the core of the
-- HerfTime Library.
-- Starting with the type 'UTCHerfTime'  which is the
-- encoding that  most other time stamps pass through

newtype UTCHerfTime = UTCHerfTime UTCTime
  deriving (Eq,Ord,Show,FormatTime)


-- | The 'ToUTCHerfTime' is necessary to have an interface 'lifted'
-- so that all the functions can be abstracted over it

class  ToUTCHerfTime a where
  herf :: a -> UTCHerfTime


-- | come back from the UTCHerfTime universal represenation
class FromUTCHerfTime a where
  unherf ::  UTCHerfTime -> a


-- | 'reherf' is not part of a typeclass, just more sugar to make dealing with time conversion
-- If both classes are defined on the same type, the classes should round trip.
-- e.g.
-- (date y m d ):: UTCTime in d == (reherf d))


reherf :: (ToUTCHerfTime a, ToUTCHerfTime b, FromUTCHerfTime a, FromUTCHerfTime b ) =>
          (a -> b)
reherf = unherf.herf


-- | This defines the time language of herf
-- the important rule here is path independence
-- (unherf $ (herf a) `add` i )  == (a `add` i)
--
-- This ensures that regardless of how you get to a time
-- the result will be the same
--
-- Something to notice is that rule still allows for lossy
-- Time Stamps.  The loss just has to be captured uniformally
-- in the transforms and the interval arithmetic

class (ToUTCHerfTime a, FromUTCHerfTime a) => HerfedTime a where
  addYear :: a -> HerfYear -> a
  addMonth :: a -> HerfMonth -> a
  addWeek :: a -> HerfWeek -> a
  addDay :: a -> HerfDay -> a
  addHour :: a -> HerfHour -> a
  addMinute :: a -> HerfMin -> a
  addSecond :: a -> HerfSec -> a
  addPicosecond :: a -> HerfPico -> a
  date :: HerfYear -> HerfMonth -> HerfDay -> a
  dateTime :: HerfYear -> HerfMonth -> HerfDay -> HerfHour -> HerfMin -> HerfSec ->  a
  dateTimePico :: HerfYear -> HerfMonth -> HerfDay -> HerfHour -> HerfMin -> HerfSec -> HerfPico ->  a


-- | This provides the unified interface to herfed times.
-- All instances of this class should obey the rule
-- that if some interval you are adding is equivalent in another
-- HerfAdd time then there should be no difference between adding that
-- many to a HerfedTime
-- i.e. (HerfMin 1) == (HerfSec 60)
-- (herf $ add someTime (HerfMin 1) ) == (herf $ add someTime (HerfMin 60))
class HerfAdd a where
  add :: (HerfedTime t) => t -> a -> t


instance HerfAdd HerfYear where
  add =  addYear


instance HerfAdd HerfMonth where
  add =  addMonth


instance HerfAdd HerfWeek where
  add =  addWeek


instance HerfAdd HerfDay where
  add =  addDay


instance HerfAdd HerfHour where
  add =  addHour


instance HerfAdd HerfMin where
  add =  addMinute


instance HerfAdd HerfSec where
  add =  addSecond


instance HerfAdd HerfPico where
  add =  addPicosecond


-- | Simple representation of a time interval of a year
newtype HerfYear   = HerfYear  Integer
  deriving (Num,Eq,Ord,Show)

-- | Simple representation of a time interval of a month
newtype HerfMonth  = HerfMonth Integer
  deriving (Num,Eq,Ord,Show)

-- | Simple representation of a time interval of a week
newtype HerfWeek = HerfWeek Integer
  deriving (Num,Eq,Ord,Show)

-- | Simple representation of a time interval of a day
newtype HerfDay    = HerfDay   Integer
  deriving (Num,Eq,Ord,Show)

-- | Simple representation of a time interval of a hour
newtype HerfHour   = HerfHour  Integer
  deriving (Num,Eq,Ord,Show)

-- | Simple representation of a time interval of a minute
newtype HerfMin    = HerfMin   Integer
  deriving (Num,Eq,Ord,Show)

-- | Simple representation of a time interval of a second
newtype HerfSec    = HerfSec   Integer
  deriving (Num,Eq,Ord,Show)

-- | Simple representation of a time interval of a picosecond
newtype HerfPico   = HerfPico  Integer -- Herf uses nano but whatev
  deriving (Num, Eq, Ord, Show)


-- | NominalHerf time will obey the laws of HerfedTime by default
instance ToUTCHerfTime UTCHerfTime where
  herf = id

instance FromUTCHerfTime UTCHerfTime where
  unherf  = id

instance HerfedTime UTCHerfTime where
  addYear (UTCHerfTime k) y = UTCHerfTime $ addYear k y
  addMonth (UTCHerfTime k) m = UTCHerfTime $ addMonth k m
  addWeek (UTCHerfTime k) w = UTCHerfTime $ addWeek k w
  addDay (UTCHerfTime k) d = UTCHerfTime $ addDay k d
  addHour (UTCHerfTime k) h = UTCHerfTime $ addHour k h
  addMinute (UTCHerfTime k) i = UTCHerfTime $ addMinute k i
  addSecond (UTCHerfTime k) s = UTCHerfTime $ addSecond k s
  addPicosecond (UTCHerfTime k) p = UTCHerfTime $ addPicosecond k p
  date (HerfYear y) (HerfMonth m) (HerfDay d) = dateHerf y m d
  dateTime (HerfYear y) (HerfMonth m) (HerfDay d) (HerfHour h) (HerfMin i) (HerfSec s ) = dateTimeHerf y m d h i s
  dateTimePico (HerfYear y) (HerfMonth m) (HerfDay d) (HerfHour h) (HerfMin i) (HerfSec s ) (HerfPico p ) = dateTimePicoHerf y m d h i s p
-- | UTCTime is the underlying and most important HerfTime thing
instance ToUTCHerfTime UTCTime where
  herf = UTCHerfTime



instance FromUTCHerfTime UTCTime where
  unherf (UTCHerfTime u) = u

-- | Get Times in any viable format (UTC for example)
-- >>> unherf $ date 2016 01 01 `add` hour 3 `add` week 16 `add` month 3   :: UTCTime
-- 2016-07-22 03:00:00 UTC


instance HerfedTime UTCTime where
  addYear (UTCTime d t) (HerfYear y) = UTCTime (addGregorianYearsRollOver y d) t
  addMonth (UTCTime d t) (HerfMonth m) = UTCTime (addGregorianMonthsRollOver m d) t
  addWeek  (UTCTime d t) (HerfWeek w) = UTCTime (addDays (7*w) d) t
  addDay (UTCTime d t) (HerfDay ds) = UTCTime (addDays ds d) t
  addHour u (HerfHour h) = addUTCTime (fromIntegral $ h*3600) u
  addMinute u (HerfMin i) = addUTCTime (fromIntegral $ i*60) u
  addSecond u (HerfSec s) = addUTCTime (fromIntegral s) u
  addPicosecond u (HerfPico p) = addUTCTime (toNominal p) u
    where
      toNominal = fromRational . toRational .  picosecondsToDiffTime

  date (HerfYear y) (HerfMonth m) (HerfDay d) = unherf $ dateHerf y m d
  dateTime (HerfYear y) (HerfMonth m) (HerfDay d) (HerfHour h) (HerfMin i) (HerfSec s ) = unherf $ dateTimeHerf y m d h i s
  dateTimePico (HerfYear y) (HerfMonth m) (HerfDay d) (HerfHour h) (HerfMin i) (HerfSec s ) (HerfPico p ) = unherf $ dateTimePicoHerf y m d h i s p


















--------------------------------------------------
-- Retrieval
--------------------------------------------------
-- | getYear someUtcTime -> 2016y
getYear :: UTCTime -> Integer
getYear incomingTime = case incomingTime of
   (UTCTime d _) -> let (year',_ ,_ ) = toGregorian d
                     in year'


-- | getMonth someUtcTime -> 10m
getMonth :: UTCTime -> Integer
getMonth incomingTime = case incomingTime of
   (UTCTime d _) -> let (_ , month' ,_ ) = toGregorian d
                    in  fromIntegral month'


-- | getDay someUtcTime -> 31d
getDay :: UTCTime -> Integer
getDay incomingTime = case incomingTime of
   (UTCTime d _) -> let (_ , _ ,day' ) = toGregorian d
                    in  fromIntegral day'


-- | Get all date parts together
getDateParts :: UTCTime -> (Integer,Integer,Integer)
getDateParts (UTCTime d _)  = (year',fromIntegral month',fromIntegral day')
  where
    (year',month',day') = toGregorian d

-- | getHour someUtcTime -> 1h
getHour :: UTCTime -> Integer
getHour (UTCTime _ t) = floor (t / 3600)

-- | getMin someUtcTime -> 37i
getMin :: UTCTime -> Integer
getMin u@(UTCTime _ t) = div remainingSeconds 60
  where
    timeInSeconds = floor t
    remainingSeconds = timeInSeconds - secondsInHours
    secondsInHours = 3600 * (getHour u)


-- | getSeconds someUtcTime -> 1s
getSeconds :: UTCTime -> Integer
getSeconds u@(UTCTime _ t) = remainingSeconds
  where
    timeInSeconds = floor t
    remainingSeconds = timeInSeconds - secondsInHours - secondsInMinutes
    secondsInHours = 3600 * (getHour u)
    secondsInMinutes = 60 * (getMin u)


-- | getPicoseconds somUTCTime -> 1ps
getPicoseconds :: UTCTime -> Integer
getPicoseconds u@(UTCTime _ t) = round $ remainingPico *
                                  (fromRational (10^(12 :: Integer)))
  where
    remainingPico = t - (fromIntegral $ secondsInHours - secondsInMinutes - seconds')
    secondsInHours = 3600 * (getHour u)
    secondsInMinutes = 60 * (getMin u)
    seconds' =  getSeconds u
